README.md
# Flash Chat ⚡️

A global chatroom for all my friend-circles brought together anonymously 🔯

## Contributions

- everyone who tries FlashChat is a contributor 🌟
- we're grateful for your bug reports and complaints 🙏
- your PRs are always welcome (kindly excuse us if it
  takes us a while to review or if there are questions
  and concerns preventing its merging).

## backend tools

- Firebase Cloud Firestore
- Firebase authentication package (cloud-based NoSQL database and secure authentication methods)
- Flutter Hero widget (for animations)
- Flutter's animation controller

**technical features used**

- Firebase integration with flutter projects
- Firebase authentication
- mixins
- Streams
- ListViews to build scrolling views
- Firebase Could Firestore to store and retrieve data

![Login](https://github.com/kamalx/images/blob/main/FlashChat1.gif)
![Chat Screen](https://github.com/kamalx/images/blob/main/FlashChat2.gif)

>This is a companion project to The App Brewery's Complete Flutter Development Bootcamp, check out the full course at [www.appbrewery.co](https://www.appbrewery.co/)


CHANGELOG.md
0.0.3
===
- fixed the single digit minutes bug in timestamps
- fixed the missing notification for login error (wrong password only)
- fixed persistent spinner even after server response was received,
  with no input to the user. user can now retry login if they entered
  wrong password without restarting the app.
- learned how to use a SnackBar and Scaffold.of()
  yes, go here: https://api.flutter.dev/flutter/material/Scaffold/of.html


0.0.2
===
- lots of bug fixes
- lots of visual ui changes
- messages now have timestamps
- best version of FlashChat ever! :tada:


0.0.1
===
- chats now actually work!
- animated welcome screen (:P)
- hero transition animations (:P)


lib/main.dart
import 'package:firebase_auth/firebase_auth.dart' as FireAuth;
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:flash_chat/screens/welcome_screen.dart';
import 'package:flash_chat/screens/login_screen.dart';
import 'package:flash_chat/screens/registration_screen.dart';
import 'package:flash_chat/screens/chat_screen.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp();
  runApp(FlashChat());
}

class FlashChat extends StatelessWidget {
  final _auth = FireAuth.FirebaseAuth.instance;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'FlashChat',
      initialRoute:
          (_auth.currentUser != null) ? ChatScreen.id : WelcomeScreen.id,
      routes: {
        WelcomeScreen.id: (context) => WelcomeScreen(),
        LoginScreen.id: (context) => LoginScreen(),
        RegistrationScreen.id: (context) => RegistrationScreen(),
        ChatScreen.id: (context) => ChatScreen()
      },
    );
  }
}


lib/constants.dart
import 'package:flutter/material.dart';

const kSendButtonTextStyle = TextStyle(
  color: Color(0xFF0277BD), // lightBlue[800]
  fontWeight: FontWeight.bold,
  fontSize: 18.0,
);

const kMessageTextFieldDecoration = InputDecoration(
  contentPadding: EdgeInsets.symmetric(vertical: 10.0, horizontal: 20.0),
  hintText: 'Type your message here...',
  border: InputBorder.none,
);

const kMessageContainerDecoration = BoxDecoration(
  border: Border(
    top: BorderSide(color: Colors.lightBlueAccent, width: 2.0),
  ),
);

const kTextFieldDecoration = InputDecoration(
  hintText: 'Enter details',
  hintStyle: TextStyle(color: Colors.black54),
  contentPadding: EdgeInsets.symmetric(vertical: 10.0, horizontal: 20.0),
  border: OutlineInputBorder(
    borderRadius: BorderRadius.all(Radius.circular(32.0)),
  ),
  enabledBorder: OutlineInputBorder(
    borderSide: BorderSide(color: Colors.lightBlueAccent, width: 1.0),
    borderRadius: BorderRadius.all(Radius.circular(32.0)),
  ),
  focusedBorder: OutlineInputBorder(
    borderSide: BorderSide(color: Colors.lightGreen, width: 1.0),
    borderRadius: BorderRadius.all(Radius.circular(32.0)),
  ),
);


lib/screens/login_screen.dart
import 'package:firebase_auth/firebase_auth.dart' as FireAuth;
import 'package:flash_chat/components/rounded_button.dart';
import 'package:flash_chat/constants.dart';
import 'package:flash_chat/screens/chat_screen.dart';
import 'package:flutter/material.dart';
import 'package:modal_progress_hud/modal_progress_hud.dart';

class LoginScreen extends StatefulWidget {
  static const String id = 'login_screen';

  @override
  _LoginScreenState createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {
  final _auth = FireAuth.FirebaseAuth.instance;
  String email, password;
  bool _showspinner = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: Builder(
        builder: (BuildContext context) {
          return ModalProgressHUD(
            inAsyncCall: _showspinner,
            child: Padding(
              padding: EdgeInsets.symmetric(horizontal: 24.0),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: <Widget>[
                  Flexible(
                    child: Hero(
                      tag: 'logo',
                      child: Container(
                        height: 200.0,
                        child: Image.asset('images/logo.png'),
                      ),
                    ),
                  ),
                  SizedBox(
                    height: 48.0,
                  ),
                  TextField(
                    textAlign: TextAlign.center,
                    keyboardType: TextInputType.emailAddress,
                    onChanged: (value) {
                      email = value;
                    },
                    style: TextStyle(color: Colors.black),
                    decoration: kTextFieldDecoration.copyWith(
                      hintText: 'Enter your email',
                    ),
                  ),
                  SizedBox(
                    height: 8.0,
                  ),
                  TextField(
                    textAlign: TextAlign.center,
                    obscureText: true,
                    onChanged: (value) {
                      password = value;
                    },
                    style: TextStyle(color: Colors.black),
                    decoration: kTextFieldDecoration.copyWith(
                      hintText: 'Enter your password',
                    ),
                  ),
                  SizedBox(
                    height: 24.0,
                  ),
                  RoundedButton(
                    color: Colors.lightBlueAccent,
                    title: 'Log In',
                    onPressed: () async {
                      setState(() {
                        _showspinner = true;
                      });
                      try {
                        var session = await _auth.signInWithEmailAndPassword(
                          email: email,
                          password: password,
                        );
                        if (session != null) {
                          print(session);
                          // navigate to the chat screen
                          Navigator.pushNamed(context, ChatScreen.id);
                        }

                        setState(() {
                          _showspinner = false;
                        });
                      } on FireAuth.FirebaseAuthException catch (e) {
                        setState(() {
                          _showspinner = false;
                        });
                        if (e.code == 'wrong-password') {
                          Scaffold.of(context).showSnackBar(
                            SnackBar(
                              backgroundColor: Colors.grey[900],
                              content: Text(
                                'Wrong password',
                                textAlign: TextAlign.center,
                                style: TextStyle(fontSize: 13),
                              ),
                            ),
                          );
                          print('\n\nbad password!\n\n');
                        }
                        print('Error occurred during signin: $e');
                      } catch (e) {
                        setState(() {
                          _showspinner = false;
                        });
                        print(
                            'Exception: code: ${e.code}, message: ${e.message}');
                      }
                    },
                  ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}


lib/screens/welcome_screen.dart
import 'package:flash_chat/screens/login_screen.dart';
import 'package:flash_chat/screens/registration_screen.dart';
import 'package:flash_chat/components/rounded_button.dart';
import 'package:flutter/material.dart';
import 'package:animated_text_kit/animated_text_kit.dart';

class WelcomeScreen extends StatefulWidget {
  static const String id = 'welcome_screen';

  @override
  _WelcomeScreenState createState() => _WelcomeScreenState();
}

class _WelcomeScreenState extends State<WelcomeScreen>
    with SingleTickerProviderStateMixin {
  AnimationController controller;
  Animation animation;

  // make sure we do some cleanup on dispose()
  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();

    controller = AnimationController(
      duration: Duration(seconds: 2),
      vsync: this,
    );

    animation = ColorTween(
      begin: Colors.blueGrey[200],
      end: Colors.white,
    ).animate(controller);

    controller.forward();

    controller.addListener(() {
      setState(() {});
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: animation.value,
      body: Padding(
        padding: EdgeInsets.symmetric(horizontal: 24.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            Row(
              children: <Widget>[
                Hero(
                  tag: 'logo',
                  child: Container(
                    child: Image.asset('images/logo.png'),
                    height: 60.0,
                  ),
                ),
                Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      'Flash ',
                      style: TextStyle(
                          fontSize: 40.0,
                          fontWeight: FontWeight.w900,
                          color: Colors.grey[900]),
                    ),
                    RotateAnimatedTextKit(
                      transitionHeight: 60.0,
                      duration: Duration(milliseconds: 800),
                      pause: Duration(milliseconds: 300),
                      repeatForever: true,
                      text: ['Fun', 'Joy', 'Chat'],
                      textStyle: TextStyle(
                        fontSize: 40.0,
                        fontWeight: FontWeight.w900,
                        color: Colors.grey[900],
                      ),
                    ),
                  ],
                ),
              ],
            ),
            SizedBox(
              height: 48.0,
            ),
            RoundedButton(
              color: Colors.lightBlueAccent,
              title: 'Log In',
              onPressed: () {
                //Go to login screen.
                Navigator.pushNamed(context, LoginScreen.id);
              },
            ),
            RoundedButton(
              color: Colors.blueAccent,
              title: 'Register',
              onPressed: () {
                //Go to registration screen.
                Navigator.pushNamed(context, RegistrationScreen.id);
              },
            ),
          ],
        ),
      ),
    );
  }
}


lib/components/rounded_button.dart
import 'package:flutter/material.dart';

/// [title] and [onPressed] are required.
/// [title] is a String to be shown as button caption
/// [onPressed] is a callback to be run when button pressed
/// [color] is the color of the button, defaults to red.
///
class RoundedButton extends StatelessWidget {
  final Color color;
  final String title;
  final Function onPressed;
  final double vpad;
  final double radius;
  final double width;
  final double height;

  RoundedButton({
    @required this.title,
    @required this.onPressed,
    this.color = Colors.red,
    this.vpad = 16.0,
    this.radius = 30.0,
    this.width = 200.0,
    this.height = 42.0,
  });

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.symmetric(vertical: vpad),
      child: Material(
        elevation: 5.0,
        color: color,
        borderRadius: BorderRadius.circular(radius),
        child: MaterialButton(
          onPressed: onPressed,
          minWidth: width,
          height: height,
          child: Text(
            title,
            style: TextStyle(
              color: Colors.white,
            ),
          ),
        ),
      ),
    );
  }
}


lib/screens/registration_screen.dart
import 'package:flash_chat/components/rounded_button.dart';
import 'package:flash_chat/constants.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flash_chat/screens/chat_screen.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:modal_progress_hud/modal_progress_hud.dart';

class RegistrationScreen extends StatefulWidget {
  static const String id = 'registration_screen';

  @override
  _RegistrationScreenState createState() => _RegistrationScreenState();
}

class _RegistrationScreenState extends State<RegistrationScreen> {
  final _auth = FirebaseAuth.instance;
  String email, password;
  bool _showspinner = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: ModalProgressHUD(
        inAsyncCall: _showspinner,
        child: Padding(
          padding: EdgeInsets.symmetric(horizontal: 24.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Flexible(
                child: Hero(
                  tag: 'logo',
                  child: Container(
                    height: 200.0,
                    child: Image.asset('images/logo.png'),
                  ),
                ),
              ),
              SizedBox(
                height: 48.0,
              ),
              TextField(
                textAlign: TextAlign.center,
                keyboardType: TextInputType.emailAddress,
                onChanged: (value) {
                  email = value;
                },
                decoration: kTextFieldDecoration.copyWith(
                  hintText: 'Enter your email',
                ),
              ),
              SizedBox(
                height: 8.0,
              ),
              TextField(
                textAlign: TextAlign.center,
                obscureText: true,
                onChanged: (value) {
                  password = value;
                },
                decoration: kTextFieldDecoration.copyWith(
                  hintText: 'Enter your password',
                ),
              ),
              SizedBox(
                height: 24.0,
              ),
              RoundedButton(
                color: Colors.blueAccent,
                title: 'Register',
                onPressed: () async {
                  setState(() {
                    _showspinner = true;
                  });
                  try {
                    final newUser = await _auth.createUserWithEmailAndPassword(
                      email: email,
                      password: password,
                    );
                    if (newUser != null) {
                      Navigator.pushNamed(context, ChatScreen.id);
                    }
                    setState(() {
                      _showspinner = false;
                    });
                  } on PlatformException catch (e) {
                    print('PlatformException was thrown: ');
                    print(
                        'createUserWithEmailAndPassword failed with error : $e');
                    setState(() {
                      _showspinner = false;
                    });
                  } catch (e) {
                    setState(() {
                      _showspinner = false;
                    });
                    print('Exception: code: ${e.code}, message: ${e.message}');
                  }
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/screens/chat_screen.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flash_chat/screens/welcome_screen.dart';
import 'package:flutter/material.dart';
import 'package:flash_chat/constants.dart';
import 'package:firebase_auth/firebase_auth.dart' as FireAuth;

final _firestore = FirebaseFirestore.instance;
FireAuth.User loggedInUser;

class ChatScreen extends StatefulWidget {
  static const String id = 'chat_screen';

  @override
  _ChatScreenState createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen> {
  final TextEditingController msgTextController = TextEditingController();
  final _auth = FireAuth.FirebaseAuth.instance;
  String messageText;

  @override
  void initState() {
    super.initState();

    getCurrentUser();
  }

  void getCurrentUser() {
    try {
      final user = _auth.currentUser; // this is now a getter
      if (user != null) {
        loggedInUser = user;
        print('Logged in user: ${loggedInUser.email}');
      }
    } catch (e) {
      print(e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: null,
        automaticallyImplyLeading: false,
        actions: <Widget>[
          IconButton(
              icon: Icon(Icons.logout),
              onPressed: () {
                try {
                  _auth.signOut();
                  Navigator.popUntil(
                    context,
                    ModalRoute.withName(WelcomeScreen.id),
                  );
                } catch (e) {
                  print('something went wrong, we could not log user out: $e');
                }
              }),
        ],
        title: Text('⚡️Chat'),
        backgroundColor: Colors.lightBlue.shade800,
      ),
      body: SafeArea(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            MessageStream(),
            Container(
              decoration: kMessageContainerDecoration,
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: <Widget>[
                  Expanded(
                    child: TextField(
                      controller: msgTextController,
                      onChanged: (value) {
                        //Do something with the user input.
                        messageText = value;
                      },
                      decoration: kMessageTextFieldDecoration,
                    ),
                  ),
                  FlatButton(
                    child: Text(
                      'Send',
                      style: kSendButtonTextStyle,
                    ),
                    onPressed: () {
                      msgTextController.clear();
                      // [messageText] + [loggedInUser.email]
                      if (messageText != null && messageText.isNotEmpty) {
                        var ref = _firestore.collection('messages').add({
                          'text': messageText ?? ' ', // paranoia?
                          'sender': loggedInUser.email ?? 'anonymous',
                          'time': Timestamp.now(), // depends on cloud_firestore
                        });
                        ref.then((value) {
                          messageText = null;
                          print('collection.add => $value');
                        });
                      } else {
                        // FIXME: disable send button
                      }
                    },
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class MessageStream extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return StreamBuilder<QuerySnapshot>(
      stream: _firestore.collection('messages').orderBy('time').snapshots(),
      builder: (context, snapshotContainer) {
        if (!snapshotContainer.hasData) {
          return Expanded(
            child: Center(
              child: CircularProgressIndicator(
                backgroundColor: Colors.lightBlue.shade800,
              ),
            ),
          );
        }
        final messages = snapshotContainer.data.docs.reversed;
        // snapshopContainer.data is a QuerySnapshot
        // final howManyMessages = messages.length;
        List<MessageBubble> messageBubbles = [];
        for (var message in messages) {
          final msgText = message.data()['text'] ??
              '[empty message]'; // this cannot happen! :P
          final msgSender = message.data()['sender'];
          final timestamp = message.data()['time'];
          final currentUser = loggedInUser.email;

          final msgBubble = MessageBubble(
            text: msgText,
            sender: msgSender,
            time: timestamp,
            isMe: currentUser == msgSender,
          );
          messageBubbles.add(msgBubble);
        }

        return Expanded(
          child: ListView(
            reverse: true,
            padding: EdgeInsets.all(10.0),
            children: messageBubbles,
          ),
        );
      },
    );
  }
}

class MessageBubble extends StatelessWidget {
  const MessageBubble({
    @required this.text,
    @required this.sender,
    @required this.time,
    this.isMe,
  });

  final String text;
  final String sender;
  final time;
  final bool isMe;

  @override
  Widget build(BuildContext context) {
    DateTime timestamp =
        DateTime.fromMillisecondsSinceEpoch(time.millisecondsSinceEpoch);

    var shape1 = BorderRadius.only(
      bottomLeft: Radius.circular(10.0),
      bottomRight: Radius.circular(30.0),
      topRight: Radius.circular(30.0),
    );
    var shape2 = BorderRadius.only(
      bottomLeft: Radius.circular(30.0),
      bottomRight: Radius.circular(10.0),
      topLeft: Radius.circular(30.0),
    );
    return Padding(
      padding: const EdgeInsets.all(1.0),
      child: Column(
        crossAxisAlignment:
            isMe ? CrossAxisAlignment.end : CrossAxisAlignment.start,
        children: [
          Material(
            borderRadius: isMe ? shape2 : shape1,
            elevation: 1.0,
            color: isMe ? Colors.lightBlue.shade800 : Colors.black12,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: [
                Padding(
                  padding: EdgeInsets.only(
                    top: 10.0,
                    left: 20.0,
                    right: 20.0,
                  ),
                  child: Text(
                    text,
                    style: TextStyle(
                      fontSize: 16.0,
                      color: isMe ? Colors.white : Colors.black,
                    ),
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(
                      bottom: 8.0, right: 20.0, left: 20.0, top: 4.0),
                  child: Text(
                    '${timestamp.hour.toString().padLeft(2, '0')}:${timestamp.minute.toString().padLeft(2, '0')}',
                    textAlign: TextAlign.end,
                    style: TextStyle(
                      color: Colors.white60,
                      fontSize: 11.0,
                      fontFamily: 'monospace',
                    ),
                  ),
                ),
              ],
            ),
          ),
          isMe
              ? SizedBox(height: 0)
              : Text(
                  sender,
                  style: TextStyle(
                    color: Colors.grey,
                    fontFamily: 'monospace',
                  ),
                ),
        ],
      ),
    );
  }
}
